include/dice/sparse_map/internal/sparse_hash.hpp

```diff
@@ -0,0 +1,943 @@
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+#ifndef DICE_SPARSE_MAP_SPARSE_HASH_HPP
+#define DICE_SPARSE_MAP_SPARSE_HASH_HPP
+
+#include <algorithm>
+#include <bit>
+#include <cassert>
+#include <cmath>
+#include <cstddef>
+#include <cstdint>
+#include <iterator>
+#include <limits>
+#include <memory>
+#include <stdexcept>
+#include <tuple>
+#include <type_traits>
+#include <utility>
+
+#include "sparse_bucket.hpp"
+#include "sparse_bucket_array.hpp"
+#include "../sparse_growth_policy.hpp"
+
+namespace dice::sparse_map::internal {
+	template<typename U>
+	struct is_power_of_two_policy : std::false_type {
+	};
+
+	template<std::size_t GrowthFactor>
+	struct is_power_of_two_policy<power_of_two_growth_policy<GrowthFactor>> : std::true_type {
+	};
+
+	/**
+	 * Internal common class used by `sparse_map` and `sparse_set`.
+	 *
+	 * `ValueType` is what will be stored by `sparse_hash` (usually `std::pair<Key,
+	 * T>` for map and `Key` for set).
+	 *
+	 * `k_select` should be a `FunctionObject` which takes a `ValueType` in
+	 * parameter and returns a reference to the key.
+	 *
+	 * `ValueSelect` should be a `FunctionObject` which takes a `ValueType` in
+	 * parameter and returns a reference to the value. `ValueSelect` should be void
+	 * if there is no value (in a set for example).
+	 *
+	 * The strong exception guarantee only holds if `ExceptionSafety` is set to
+	 * `dice::sh::exception_safety::strong`.
+	 *
+	 * `ValueType` must be nothrow move constructible and/or copy constructible.
+	 * Behaviour is undefined if the destructor of `ValueType` throws.
+	 *
+	 *
+	 * The class holds its buckets in a 2-dimensional fashion. Instead of having a
+	 * linear `std::vector<bucket>` for [0, bucket_count) where each bucket stores
+	 * one value, we have a `std::vector<sparse_array_type>` (buckets_)
+	 * where each `sparse_array_type` stores multiple values (up to
+	 * `sparse_array_type::discriminant_bits`). To convert a one dimensional `ibucket`
+	 * position to a position in `std::vector<sparse_array_type>` and a position in
+	 * `sparse_array_type`, use respectively the methods
+	 * `sparse_array_type::sparse_ibucket(ibucket)` and
+	 * `sparse_array_type::index_in_sparse_bucket(ibucket)`.
+	 */
+	template<typename ValueType,
+			 typename KeyValueSelect,
+			 typename Hash,
+			 typename KeyEqual,
+			 typename Allocator,
+			 growth_policy GrowthPolicy,
+			 exception_safety ExceptionSafety,
+			 sparsity Sparsity,
+			 probing Probing,
+			 ratio MaxLoadFactor>
+	struct sparse_hash {
+	private:
+		template<typename VSel>
+		struct get_mapped_type {
+			using type = void;
+			using const_reference = void;
+			using reference = void;
+		};
+
+		template<typename VSel> requires requires { typename VSel::value_type; }
```

Comment:
why requires requires

Reply:
Because we are requiring something using `requires ...` (= requires declaration) but `...` in our case isn't a named concept so we create one on the fly by utilizing `requires { ... }` (= requires expression).

Kind of simmilar to `void f() noexcept(noexcept(expression))`

Reply:
You can also do `static_assert(requires { ... })` btw if that makes it clearer.